iroh/src/node.rs

```diff
@@ -782,21 +783,85 @@ impl<D: BaoStore> RpcHandler<D> {
+        use bao_tree::io::fsm::Outboard;
+
+        let db = self.inner.db.clone();
+        for blob in db.blobs()? {
+            let blob = blob?;
+            let Some(entry) = db.get(&blob)? else {
+                continue;
+            };
+            let hash = entry.hash();
+            let size = entry.outboard().await?.tree().size().0;
+            let path = "".to_owned();
+            co.yield_(BlobListResponse::Item { hash, size, path }).await;
+        }
+        Ok(())
+    }
+
+    async fn blob_list_incomplete_impl(
+        self,
+        co: &Co<BlobListIncompleteResponse>,
+    ) -> io::Result<()> {
+        let db = self.inner.db.clone();
+        for hash in db.partial_blobs()? {
+            let hash = hash?;
+            let Ok(PossiblyPartialEntry::Partial(entry)) = db.get_possibly_partial(&hash) else {
+                continue;
+            };
+            let size = 0;
+            let expected_size = entry.size();
+            co.yield_(BlobListIncompleteResponse::Item {
+                hash,
+                size,
+                expected_size,
+            })
+            .await;
+        }
+        Ok(())
+    }
+
+    async fn blob_list_collections_impl(
+        self,
+        co: &Co<BlobListCollectionsResponse>,
+    ) -> anyhow::Result<()> {
+        let db = self.inner.db.clone();
+        let local = self.inner.rt.clone();
+        let tags = db.tags().unwrap();
+        for item in tags {
+            let (name, HashAndFormat { hash, format }) = item?;
+            if !format.is_hash_seq() {
+                continue;
+            }
+            let Some(entry) = db.get(&hash)? else {
+                continue;
+            };
+            let count = local
+                .spawn_pinned(|| async move {
+                    let reader = entry.data_reader().await?;
+                    let (_collection, count) = parse_hash_seq(reader).await?;
+                    anyhow::Ok(count)
+                })
+                .await??;
+            co.yield_(BlobListCollectionsResponse::Item {
+                tag: name,
+                hash,
+                total_blobs_count: Some(count),
+                total_blobs_size: None,
+            })
+            .await;
+        }
+        Ok(())
+    }
+
     fn blob_list(
         self,
         _msg: BlobListRequest,
     ) -> impl Stream<Item = BlobListResponse> + Send + 'static {
-        use bao_tree::io::fsm::Outboard;
-
-        let db = self.inner.db.clone();
-        futures::stream::iter(db.blobs()).filter_map(move |hash| {
-            let db = db.clone();
-            async move {
-                let entry = db.get(&hash)?;
-                let hash = entry.hash().into();
-                let size = entry.outboard().await.ok()?.tree().size().0;
-                let path = "".to_owned();
-                Some(BlobListResponse { hash, size, path })
+        Gen::new(|co| async move {
```

Comment:
I guess this pattern is so common that we might want to do a helper for it. But maybe not just yet. It isn't that long.